stan/math/prim/fun/grad_2F1.hpp

```diff
@@ -24,82 +74,126 @@ namespace math {
  * This power-series representation converges for all gradients
  * under the same conditions as the 2F1 function itself.
  *
- * @tparam T type of arguments and result
- * @param[out] g_a1 g_a1 reference to gradient of 2F1 w.r.t. a1, result.
- * @param[out] g_b1 g_b1 reference to gradient of 2F1 w.r.t. b1, result.
- * @param[in] a1 a1 see generalized hypergeometric function definition.
- * @param[in] a2 a2 see generalized hypergeometric function definition.
- * @param[in] b1 b1 see generalized hypergeometric function definition.
- * @param[in] z z see generalized hypergeometric function definition.
+ * @tparam T1 scalar or `var_value`
+ * @tparam T2 scalar or `var_value`
+ * @tparam T3 scalar or `var_value`
+ * @tparam T_z scalar type
+ * @param[out] g_a1 g_a1 reference to gradient of 2F1 w.r.t. a1, result
+ * @param[out] g_a2 g_a2 reference to gradient of 2F1 w.r.t. a2, result
+ * @param[out] g_b1 g_b1 reference to gradient of 2F1 w.r.t. b1, result
+ * @param[in] a1 a1 see generalized hypergeometric function definition
+ * @param[in] a2 a2 see generalized hypergeometric function definition
+ * @param[in] b1 b1 see generalized hypergeometric function definition
+ * @param[in] z z see generalized hypergeometric function definition
  * @param[in] precision magnitude of the increment of the infinite sum
- *   to truncate the sum at.
- * @param[in] max_steps number of steps to take.
+ *   to truncate the sum
+ * @param[in] max_steps number of steps to take
  */
-template <typename T>
-void grad_2F1(T& g_a1, T& g_b1, const T& a1, const T& a2, const T& b1,
-              const T& z, double precision = 1e-14, int max_steps = 1e6) {
+template <typename T1, typename T2, typename T3, typename T_z>
+void grad_2F1(T1& g_a1, T2& g_a2, T3& g_b1, const T1& a1, const T2& a2,
+              const T3& b1, const T_z& z, double precision = 1e-14,
+              int max_steps = 1e6) {
   check_2F1_converges("grad_2F1", a1, a2, b1, z);
 
   using stan::math::value_of_rec;
   using std::exp;
   using std::fabs;
   using std::log;
   using std::max;
+  using TP = return_type_t<T1, T2, T3, T_z>;
```

Comment:
We use `_t` to and snake case to signify a type returned by a type trait

```suggestion
  using ret_t = return_type_t<T1, T2, T3, T_z>;
```